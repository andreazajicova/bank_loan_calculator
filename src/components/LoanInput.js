import React from 'react';
import './stylesheet/LoanInput.css'

export default function LoanInput({ 
    setInterest,
    saveLoanAmount,
    currentInterest,
    saveYears,
    calculatedLoanResults,
    
}) {


    return (
        <div className="Dashboard__wrapper">
            <div className="Dashboard__calculator">
                <form className="Dashboard__form">
                    <div className="Dashboard__select">
                        <p>Select a type of loan: </p>
                            <select className="Dashboard__value" onChange={setInterest} id="select_id" defaultValue="Select">
                                <option value="select">Select</option>
                                <option value="house">House</option>
                                <option value="car">Car</option>
                                <option value="spending">Spending</option>
                                <option value="other">Other</option>
                            </select>   
                    </div>

                    <div className="Dashboard__input">
                        <p>Loan amount:  </p>
                        <input type="number" min="10000" max="10000000" placeholder="Loan amount" onChange={saveLoanAmount}/>
                        <p> kr</p>
                    </div>

                    <div className="Dashboard__interest">
                        <p>Interest: {currentInterest} %</p>
                    </div>

                    <div className="Dashboard__payback">
                        <p>Payback time:  </p>
                        <input data-testid="years" type="number" min="1" placeholder="Number of years" onChange={saveYears}/>
                        <p> years</p>
                    </div>

                    <button data-testid="submit" className="Dashboard__button" type="submit" onClick={calculatedLoanResults}>
                        Calculate
                    </button>
                </form>
            </div>  
        </div>
    )
}
