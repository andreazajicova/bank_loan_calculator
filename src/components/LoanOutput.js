import React from 'react';
import { v4 as uuid_v4 } from "uuid";

export default function LoanOutput({ 
    years,
    currentInterest,
    monthlyPayment,
    monthlyInterests

}) {

    return (
        <div className="Loan__wrapper">
            <p>Your payback time is: {years} year(s).</p>
            <p>With {currentInterest} % yearly interest the payback of your loan looks like following: </p>
            
            <div>
                <p>Your fixed payment every month: <strong>{monthlyPayment}</strong></p>
            </div>

            <div>
                <p>Your interest month by month:</p>
                { monthlyInterests && monthlyInterests.map(interest => {
                    return (
                        <div key={uuid_v4()}>
                            <strong> {interest} kr</strong>
                        </div>
                    )
                })}
            </div>
        </div>
    )
}
