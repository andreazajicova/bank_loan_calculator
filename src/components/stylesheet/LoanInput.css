/* .Dashboard__wrapper {
    border: 10px solid white;
} */

.Dashboard__wrapper {
    display: flex;
    justify-content: center;
    margin-top: 15px;
    /* border: solid 1px black;
    width: 300px;
    margin: auto;
    padding: 15px; */
}

.Dashboard__select {
    display: flex;
    flex-direction: row;
}

.Dashboard__value {
    margin-top: 15px;
    margin-left: 5px;
    height: 20px;
    width: 85px;
}

.Dashboard__input, .Dashboard__interest, .Dashboard__payback {
    display: flex;
    flex-direction: row;
}

input {
    margin-left: 5px;
    margin-right: 5px;
    margin-top: 15px;
    height: 15px;
    width: 130px;
}

.Dashboard__button {
    margin-top: 5px;
    font-size: 18px;
    margin-bottom: 70px;
}