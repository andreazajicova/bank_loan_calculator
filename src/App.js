import React, { useState } from 'react';
import Header from './components/Header';
import LoanInput from './components/LoanInput';
import LoanOutput from './components/LoanOutput';
import './App.css';

const App = () => {

  const [submitClick, setSubmitClick] = useState(false);
  const [currentInterest, setCurrentInterest] = useState(3.5);
  const [carLoanInterest, setCarLoanInterest] = useState(3.5);
  const [houseLoanInterest, setHouseLoanInterest] = useState(3.5);
  const [spendingLoanInterest, setSpendingLoanInterest] = useState(1);
  const [otherLoanInterest, setOtherLoanInterest] = useState(3.5);
  const [loanAmount, setLoanAmount] = useState(0);
  const [years, setYears] = useState(0);
  const [monthlyInterests, setMonthlyInterests] = useState([]);
  const [monthlyPayment, setMonthlyPayment] = useState(0);
    
  const saveLoanAmount = (event) => {
    const loan = +(event.target.value);
    if (loan > 10000000) {
      alert('The maximum possible loan amount is 10000000 kr.')
    } 
    setLoanAmount(loan);
  }

  const saveYears = (event) => {
    event.preventDefault();
    let years = +(event.target.value);
    if (years > 50) {
      alert('Paybacktime must be maximum 50 years.');
    }
    setYears(years);
  }

  const calculatedLoanResults = (e) => {
    e.preventDefault();
    setSubmitClick(true);
    let array = [];
    const months = years * 12;
    const monthlyPayment = Math.round((loanAmount / months) * 100) / 100;
    let balance = loanAmount - monthlyPayment;
    const firstMonhtlyInterest = Math.round((((loanAmount / 100) * currentInterest) / 12) * 100) / 100;
    array.push(firstMonhtlyInterest);
    setMonthlyPayment(monthlyPayment);

    if(balance && balance > monthlyPayment) {
      for(let i = 0; i < months; i++) {
        const yearlyInterest = Math.round((((balance / 100) * currentInterest)) * 100) / 100;
        const monthlyInterest = Math.round((yearlyInterest / 12) * 100) / 100;
        balance -= monthlyPayment;
        console.log(monthlyInterest);
        array.push(monthlyInterest);
        setMonthlyInterests(array);
        
      }
    }     
  }
  
  
  const setInterest = () => {
    const type = document.getElementById("select_id").value;
    if (type === 'house') {
      setCurrentInterest(houseLoanInterest);
    } else if (type === 'car') {
      setCurrentInterest(carLoanInterest)
    } else if (type === 'spending') {
      setCurrentInterest(spendingLoanInterest);
    } else {
      setCurrentInterest(otherLoanInterest);
    }
  }

  return (
    <div className="App">
      
        <Header />
      
        <LoanInput 
          setInterest={setInterest}
          saveLoanAmount={saveLoanAmount}
          currentInterest={currentInterest}
          saveYears={saveYears}
          calculatedLoanResults={calculatedLoanResults}
        />

      { submitClick ? 
        (<LoanOutput 
          years={years}
          monthlyInterests={monthlyInterests}
          monthlyPayment={monthlyPayment}
          currentInterest={currentInterest}
        />
        ) : (
          ''
        )
      }
    </div>
  );
}

export default App;
