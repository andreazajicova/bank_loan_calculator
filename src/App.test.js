import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import LoanInput from './components/LoanInput';
import LoanOutput from './components/LoanOutput';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import { shallow } from 'enzyme';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App></App>, div);
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<LoanInput></LoanInput>, div);
});

it('renders props correctly', () => {
  const component = renderer.create(<LoanOutput currentInterest={3.5} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('Should show h3', ()=>{
  const component = shallow(<LoanOutput/>);
  expect(component.find("p")).toHaveLength(4);  
})

//  it('<input> updates input value on onChange event', () => {
//     const component = shallow(<LoanInput />);
//     const input = component.find('#years');
//     input.props().onChange({ target: { value: 3}});
//     wrapper.find("#years").simulate("change", mockEvent);
//     expect(component.state('input')).toEqual(3);
//   });

it('payback time input field renders the correct number', () => {
  const component = render(<LoanInput />);
  fireEvent.change(component.getByTestId('years'), {
    target: {
      value: 3
    }
  });
  expect(component.getByTestId('years').value).toBe('3');
})

it('Calculate button works', () => {
  const calculatedLoanResults = jest.fn();
  const component = render(<button data-testid="submit" className="Dashboard__button" type="submit" onClick={calculatedLoanResults}></button>);
  fireEvent.click(component.getByTestId('submit'));
  expect(calculatedLoanResults).toHaveBeenCalledTimes(1);
})

it('Calculate monhtly payment works', () => {
  const loanAmount = 10000;
  const months = 36;
  const monthlyPayment = Math.round((loanAmount / months) * 100) / 100;
  expect(monthlyPayment).toBe(277.78)
})
