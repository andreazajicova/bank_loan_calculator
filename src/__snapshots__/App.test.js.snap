// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders props correctly 1`] = `
<div
  className="Loan__wrapper"
>
  <p>
    Your payback time is: 
     year(s).
  </p>
  <p>
    With 
    3.5
     % yearly interest the payback of your loan looks like following: 
  </p>
  <div>
    <p>
      Your fixed payment every month: 
      <strong />
    </p>
  </div>
  <div>
    <p>
      Your interest month by month:
    </p>
  </div>
</div>
`;
